base/channels.jl

```diff
@@ -553,6 +718,134 @@ function take_unbuffered(c::Channel{T}) where T
+    end
+    return buffered ? take_buffered(c, n, buffer) : take_unbuffered(c, n, buffer)
+end
+
+function _resize_buffer(buffer, n, resize_count, n_to_take=1)
+    # As we take elements from the channel, we may need to resize the buffer to accommodate them
+    # for small n, we resize once at the beginning. But for larger n, we resize the buffer as needed.
+    # We resize exponentially larger each time until we reach 2^16, at which point we resize by 2^16.
+    len = length(buffer)
+    needed_space = max(
+        n_to_take,
+        2^resize_count
+    )
+    resize!(buffer, min(len + needed_space, n))
+    resize_count >= 16 && return resize_count
+    return resize_count + 1
+end
+
+function _rm_unused_slots(buffer, last_idx)
+    if last_idx <= lastindex(buffer)
+        deleteat!(buffer, last_idx+1:lastindex(buffer))
+    end
+end
+
+function take_buffered(c::Channel{T}, n, buffer::AbstractVector) where {T}
+    elements_taken = 0 # number of elements taken so far
+    idx1 = firstindex(buffer)
+    target_buffer_len = min(n, c.sz_max)
+    resize_count = 1
+
+    lock(c)
+    try
+        while elements_taken < n && (isopen(c) || isready(c))
+            # wait until the channel has at least min_n elements or is full
+            while length(c.data) < target_buffer_len && isopen(c)
+                _do_is_closed(() -> wait(c.cond_take)) && break
+            end
+            # take as many elements as possible from the buffer
+            n_to_take = min(n - elements_taken, length(c.data))
+            idx_start = idx1 + elements_taken
+            idx_end = idx_start + n_to_take - 1
+            if idx_end > lastindex(buffer)
+                resize_count = _resize_buffer(buffer, n, resize_count, n_to_take)
+            end
+            # since idx_start/end are both created relative to `firstindex(buffer)`, they are safe to use
+            # as indices for the buffer
+            @views copy!(buffer[idx_start:idx_end], c.data[1:n_to_take])
+            deleteat!(c.data, 1:n_to_take)
+            _increment_n_avail(c, -n_to_take)
+            foreach(_ -> notify(c.cond_put, nothing, true, false), 1:n_to_take)
+            elements_taken += n_to_take
+        end
+
+        _rm_unused_slots(buffer, firstindex(buffer)+elements_taken-1)
+        return buffer
+    finally
+        unlock(c)
+    end
+end
+
+function take_unbuffered(c::Channel, n, buffer::AbstractVector)
+    idx1 = firstindex(buffer)
+    last_idx = idx1 - 1
+    resize_count = 1
+    lock(c)
+    try
+        for i in idx1:(idx1+n-1)
+            if i > lastindex(buffer)
+                resize_count = _resize_buffer(buffer, n, resize_count)
+            end
+            _do_is_closed() do
+                buffer[i] = take_unbuffered(c)
+            end && break
+            last_idx = i
+        end
+
+        _rm_unused_slots(buffer, last_idx)
+        return buffer
+    finally
+        unlock(c)
+    end
+end
+
+collect(c::Channel) = take!(c, typemax(UInt))
```

Comment:
Currently, `collect(::Channel)` falls back to a generic method which just iterates over the channel. I added a custom method for `collect` that uses `take!(ch,n)` since we can benefit from maintaining the lock.

However, this would change the behavior of parallel `collect` calls on the same channel. I think multiple `collect` calls would be ill-advised anyway, but I did want to highlight that this would technically be a subtle change.